File: problems/09/9.c

// A Pythagorean triplet is a set of three natural numbers, a < b < c, for
// which,
//
// a² + b² = c²
//
// For example, 3³ + 4² = 9 + 16 = 25 = 5²
//
// There exists exactly one Pythagorean triplet for which a + b + c = 1000.
// Find the product abc.


#include <stdio.h>
#include <time.h>


int solve1() {
    for(int c = 5; c < 500; c++)
        for(int b = 1; b < c; b++)
            for(int a = 1; a < b; a++)
                if (a + b + c == 1000 && a*a + b*b == c*c)
                    return a * b *c;
}

int solve2() {
    for(int b = 5; b < 500; b++)
        for(int a = 1; a < b; a++) {
#define C (1000 - a - b)
            if(a*a + b*b == C*C)
                return a * b * C;
        }
#undef C
}


int main() {
    clock_t begin = clock();
    int result = solve1();
    clock_t end = clock();
    double time = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("%d [in %f seconds]\n", result, time);

    clock_t begin2 = clock();
    int result2 = solve2();
    clock_t end2 = clock();
    double time2 = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("%d [in %f seconds]\n", result2, time2);
}

File: problems/09/9.py
# coding: utf-8

"""
A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a² + b² = c²

For example, 3³ + 4² = 9 + 16 = 25 = 5².

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.
"""

import time


def oneliner():

    return [(a, b, c, a * b * c)
            for c in range(5, 500)
            for b in range(1, c)
            for a in range(1, b)
            if a + b + c == 1000 and a**2 + b**2 == c**2][0]


def solve1():

    for c in range(5, 500):
        for b in range(1, c):
            for a in range(1, b):
                if a + b + c == 1000 and a**2 + b**2 == c**2:
                    return a, b, c, a * b * c


def solve2():

    for b in range(5, 500):
        for a in range(1, b):
            c = 1000 - (a + b)
            if a**2 + b**2 == c**2:
                return a, b, c, a * b * c


if __name__ == "__main__":

    def timeit(function):
        t1 = time.time()
        output = function()
        t2 = time.time()
        return output, t2-t1

    print timeit(solve1)
    print timeit(solve2)
    print timeit(oneliner)
